iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java

```diff
@@ -0,0 +1,202 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.queryengine.plan.relational.planner.iterative.rule;
+
+import org.apache.iotdb.db.queryengine.plan.planner.plan.node.PlanNode;
+import org.apache.iotdb.db.queryengine.plan.relational.planner.Symbol;
+import org.apache.iotdb.db.queryengine.plan.relational.planner.iterative.Rule;
+import org.apache.iotdb.db.queryengine.plan.relational.planner.node.AggregationNode;
+import org.apache.iotdb.db.queryengine.plan.relational.planner.node.MarkDistinctNode;
+import org.apache.iotdb.db.queryengine.plan.relational.utils.matching.Captures;
+import org.apache.iotdb.db.queryengine.plan.relational.utils.matching.Pattern;
+
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toSet;
+import static org.apache.iotdb.db.queryengine.plan.relational.planner.node.Patterns.aggregation;
+import static org.apache.tsfile.read.common.type.BooleanType.BOOLEAN;
+
+/**
+ * Implements distinct aggregations with different inputs by transforming plans of the following
+ * shape:
+ *
+ * <pre>
+ * - Aggregation
+ *        GROUP BY (k)
+ *        F1(DISTINCT a0, a1, ...)
+ *        F2(DISTINCT b0, b1, ...)
+ *        F3(c0, c1, ...)
+ *     - X
+ * </pre>
+ *
+ * into
+ *
+ * <pre>
+ * - Aggregation
+ *        GROUP BY (k)
+ *        F1(a0, a1, ...) mask ($0)
+ *        F2(b0, b1, ...) mask ($1)
+ *        F3(c0, c1, ...)
+ *     - MarkDistinct (k, a0, a1, ...) -> $0
+ *          - MarkDistinct (k, b0, b1, ...) -> $1
+ *              - X
+ * </pre>
+ */
+public class MultipleDistinctAggregationToMarkDistinct implements Rule<AggregationNode> {
+  private static final Pattern<AggregationNode> PATTERN =
+      aggregation()
+          .matching(
+              Predicates.and(
+                  MultipleDistinctAggregationToMarkDistinct::hasNoDistinctWithFilterOrMask,
+                  Predicates.or(
+                      MultipleDistinctAggregationToMarkDistinct::hasMultipleDistincts,
+                      MultipleDistinctAggregationToMarkDistinct::hasMixedDistinctAndNonDistincts)));
+
+  private static boolean hasNoDistinctWithFilterOrMask(AggregationNode aggregationNode) {
+    return aggregationNode.getAggregations().values().stream()
+        .noneMatch(
+            aggregation ->
+                aggregation.isDistinct()
+                    && (aggregation.getFilter().isPresent() || aggregation.getMask().isPresent()));
```

Comment:
What is the difference between filter and mask? Perhaps you need to deal with filter Using the Rule `ImplementFilteredAggregations`

Reply:
Filter is an another feature of AggFunction, it hasn't been supported in our project, so `aggregation.getFilter().isPresent()` will always be false. 
<img width="683" alt="image" src="https://github.com/user-attachments/assets/7d8bc07a-1328-4985-a7a6-c27e6d7ad1ae" />

Reply:
Will filter in Aggregation be Introduced in the future?

Reply:
may be, hhhhhh, in far away future.